doc/03_reference/coverage_plan.rst

```diff
@@ -0,0 +1,299 @@
+Each of the points listed below must be covered.
+This will be further combined using cross coverage which is described in the section below.
+
+Instructions
+^^^^^^^^^^^^
+
+Categories
+""""""""""
+Instructions can be grouped into a number of categories.
+Each category represents a different kind of execution.
+For example the ``ADD`` and ``SUB`` instructions are in the same category as they are almost identical for the microarchitecture (both read two registers and write to one, both feed operands to the ALU and take their result from it, both have the same response to interrupts etc, the only difference is the ALU operation).
+
+Instructions can be compressed or uncompressed but that isn’t factored into the instruction categories below (excepting for illegal instructions).
+The decompress occurs in the IF stage and is invisible to the ID/EX stage so isn’t relevant for instruction execution.
+A separate set of category agnostic compressed instruction behaviour is considered instead.
+
+An instruction category is sampled at the ID/EX stage (which is where all the varying behaviours actually occur).
+Some categories are just a single instruction which is named without further description.
+
+
+* ALU - All of the reg/reg reg/imm instructions that use the ALU.
+  This is any RISC-V instruction with an opcode of ``7’b0010011`` or ``7’b0110011`` (``ibex_pkg::OPCODE_OP`` and ``ibex_pkg::OPCODE_OP_IMM``) other than the ``MUL*`` and ``DIV*`` family of instructions (from RV32M).
+* Mul - Any ``MUL*`` instruction (from RV32M).
+* Div - Any ``DIV*`` instruction (from RV32M).
+* Branch - Any ``B*`` family branch instruction.
+* Jump - ``JAL``/``JALR``
+* Load - Any ``L*`` family load instruction.
+* Store - Any ``S*`` family load instruction.
+* CSRAccess - Any instruction from Zicsr.
+* EBreakDbg/EBreakExc- An ``EBREAK`` instruction that either enters debug mode (Dbg) or causes an exception (Exc).
+  Which occurs depends upon the setting of ``dcsr.ebreakm`` / ``dcsr.ebreaku`` combined with the privilege level of executed instruction.
+* ``ECALL``
+* ``MRET``
+* ``DRET``
+* ``WFI``
+* ``FENCE.I``
+* FetchError - Any instruction that saw a fetch error.
+* CompressedIllegal - Any compressed instruction with an illegal encoding.
+* UncompressedIllegal - Any uncompressed instruction with an illegal encoding.
+* CSRIllegal - Any instruction attempting a CSR access that is not allowed.
+* PrivIllegal - Illegal due to privilege level or being in/out of debug mode.
+* OtherIllegal - Any other instruction that raises an Illegal instruction exception that isn’t in the other categories.
+* None - No instruction in ID/EX stage.
+
+Stalls
+""""""
+Not all instructions can see all kinds of stalls.
+A stall category is sampled at the ID/EX stage only (as stalls in IF and WB don’t break down into categories).
+
+* Instr - A stall cause by a multi-cycle instruction.
+  This can be seen by instructions from categories:
+
+  * MUL
+  * DIV
+  * Branch
+  * Jump
```

Comment:
I wonder whether these lists are actually needed. Presumably we'll have `illegal_bins` lines somewhere: maybe that's a better place to put "hey, this can't happen for any other categories"?

Reply:
I'd prefer to leave them in for now as this will be the document used to finish off the coverage implementation so I don't want to accidentally loose anything. Once the implementation is done we can consider removing this (e.g. if we think comments in the coverage implementation itself are sufficient).